src/interpreter.ts

```diff
@@ -25,3 +480,886 @@ export function parseAndEvalExpression(sourceCode: string): EvalResult {
         throw error;
     }
 }
+
+const defaultInterpreterConfig: InterpreterConfig = {
+    // We set the default max number of loop iterations
+    // to the maximum number allowed for repeat loops
+    maxLoopIterations: maxRepeatStatement,
+};
+
+export class Interpreter {
+    private currentEnv: RuntimeEnvironment;
+    private initialContext: CompilerContext;
+    private config: InterpreterConfig;
+
+    constructor(
+        initialContext: CompilerContext = new CompilerContext(),
+        config: InterpreterConfig = defaultInterpreterConfig,
+    ) {
+        this.currentEnv = new RuntimeEnvironment();
+        this.initialContext = initialContext;
+        this.config = config;
+    }
+
+    // Opens a new environment with the provided param names and
+    // values.
+    private openNewEnvironment(
```

Comment:
Moved the opening and closing of environments into `RuntimeEnvironment` (which I renamed into `EnvironmentStack`, because that is what actually is. The stack is now a linked list of environments).
The opening and closing of environments is now done automatically inside the `EnvironmentStack` by calling `executeInNewEnvironment` method. This removes the potential error of a programmer           
forgetting to open and close environments manually. The code also looks cleaner, i.e., instead of:
```
openNewEnvironment();
try {
   // execute some code
} finally {
  closeCurrentEnvironment()
}
```
it is now:
```
envStack.executeInNewEnvironment(
 () => {
     // execute some code
  }
);
```